docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,217 @@
+	// Reap returns the next available transaction in the mempool, returning an error
+	// upon failure. The notion of 'available' or 'next' is defined by the application's
+	// mempool implementation.
+	Reap() (T, error)
+	// ReapMaxBytes performs the the same function as Reap, except it returns
+	// multiple transactions limited by their total size.
+	ReapMaxBytes(n int) ([]T, error)
+	// NumTxs returns the number of transactions currently in the mempool.
+	NumTxs() int
+}
+```
+
+We will define an implementation of `Mempool[T MempoolTx]` that will cover a
+majority of application use cases. Namely, it will prioritize transactions by
+priority and transaction sender, allowing for multiple prioritized transactions
+from the same sender. The mempool will be defined as a wrapper around a simple
+priority queue using a max binary heap, along with additional indexes/metadata
+to store senders and their nonces, allowing for simple multi-dimensional
+prioritization (2-ary).
+
+Transaction reaping will essentially happen via a two-phase approach:
+
+1. Reap one or more transactions from the priority queue and collect them into 
+   one of two buffers -- _valid_ and _invalid_.
+2. For transactions that DO NOT violate the nonce validation, they are included
+   in the _valid_ buffer.
+3. For transactions that DO violate the nonce validation, they are included in
+   the _invalid_ buffer.
+4. Continue this process until the desired number of valid transactions are
+   reaped or until the mempool is empty.
+5. Provide Tendermint the list of all transactions from the _valid_ buffer.
+6. Re-insert all transactions, from both buffers, back into mempool. This is to
+   ensure we do not discard transactions from the mempool in case `ProcessProposal`
+   fails.
+
+```go
+type PriorityMempool[T MempoolTx] struct {
+	queue   *PriorityQueue[MempoolTx]
+	senders map[string][]int64
+	// ...
+}
+```
+
+Previous discussions<sup>1</sup> have come to the agreement that Tendermint will
+perform a request to the application, via `PrepareProposalRequest`, with a certain
+amount of transactions reaped from Tendermint's local mempool. The exact amount
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
+
+When Tendermint reaps transactions from the local mempool and sends them to the
+application via `PrepareProposalRequest`, the application will have to evaluate
+the transactions. Specifically, it will need to inform Tendermint if it should
+reject and or include each transaction. Note, the application can even _replace_
+transactions entirely with other transactions.
+
+When evaluating transactions from `PrepareProposalRequest`, the application will
+ignore _all_ transactions sent to it in the request and instead reap up to
+`PrepareProposalRequest.max_tx_bytes` from it's own mempool. There is no need to
+execute the transactions for validity as they have already passed CheckTx.
```

Comment:
For the version of `RequestPrepareProposal` in v0.37, Tendermint expects the ABCI application to examine the list of Tx that was sent to it. Specifically, it requires that transactions that were _not_ part of the original list passed to the application in `RequestPrepareProposal` be marked as `ADDED` and it requires that transactions that were passed in the original list that are subsequently included in the `ResponsePrepareProposal` be marked as `UNMODIFIED`. This means that, at present, the original list has to inspected and can't just be ignored. 

I've opened https://github.com/tendermint/tendermint/issues/9225 to track future improvements that may be made to the ABCI++ that will enable the API to better support app-side mempools but at the moment, some extra logic will be required in the SDK here.

Reply:
This is not what I understood from Sergio -- from our convos, it was stated that txs can be completely omitted from the original list.

In any case, I've updated `ReapMaxBytes` per @prettymuchbryce's suggestion which now includes an `Action` for each record returned. It also accepts the original list of records so the mempool can do all the filtering and decision making it needs.

Reply:
> This is not what I understood from Sergio -- from our convos, it was stated that txs can be completely omitted from the original list.

Transactions can be completely omitted from the original list. That's definitely true. The concern is what happens if the app-side mempool returns a transaction to Tendermint in `ResponsePrepareProposal` that was present in the list handed to the application during `RequestPrepareProposal`. If that happens, the app-side mempool will need to mark the transaction as `UNMODIFIED` not `ADDED`. This means that the SDK will either need to query the list passed to it for each of the transactions it's going to return so it can determine what status to mark them with, _or_ it will need to modify the transactions somehow so they don't match the transactions originally passed to the application, perhaps by appending a null-byte or similar. This second option will also require the app to clear them all in `RecheckTx`, although, I believe that's the intention already.

Reply:
I am sorry to realize that there was some misunderstanding in our conversations: @williambanfield's description above looks fully accurate to me.
I agree the "ergonomics" of this interface can be improved for the "app-side mempool" use case. But bear in mind other Apps will instead examine the TX list received from Tendermint, which is a different use case but equally valid.
I am sure we will find a way forward here in an orderly way.

Reply:
shouldn't tendermint only care if it needs to be removed from the mempool or if it was modified? Why does it care about added or unmodified?

Reply:
>or it will need to modify the transactions somehow so they don't match the transactions originally passed to the application, perhaps by appending a null-byte or similar

Won't that compromise traceability as described in [the spec](https://github.com/tendermint/tendermint/blob/master/spec/abci%2B%2B/abci++_methods.md#parameters-and-types-1)?

>The Application should be aware that removing and adding transactions may compromise traceability.
>>Consider the following example: the Application transforms a client-submitted transaction t1 into a second transaction t2, i.e., the Application asks Tendermint to remove t1 and add t2 to the mempool. If a client wants to eventually check what happened to t1, it will discover that t1 is neither in the mempool nor in a committed block, getting the wrong idea that t1 did not make it into a block. Note that t2 will be in a committed block, but unless the Application tracks this information, no component will be aware of it. Thus, if the Application wants traceability, it is its responsability to support it. For instance, the Application could attach to a transformed transaction a list with the hashes of the transactions it derives from.

Reply:
>shouldn't tendermint only care if it needs to be removed from the mempool or if it was modified? Why does it care about added or unmodified?

There is no `MODIFIED` status, the app can only `ADD` and `REMOVE`.  `ADD` will include the tx in subsequent ABCI method calls, and possibly add the tx to the mempool.  The docs are contradictory on the second point though.

Reply:
To clarify for each `PrepareProposalResponse`:

* Tx records that are removed from the mempool due to eviction will be added to the end of the list with action `REMOVED`.
* Tx records that are to be included in the block that are NOT in the original list (i.e. the request), will have status `ADDED`.
* Tx records that are to be included in the block that are in the original list (i.e. the request), will have status `UNMODIFIED`.

Let me know if this is the consensus and I will amend the ADR to state this.

Reply:
> The docs are contradictory on the second point though.

I confirm that the v0.36.x spec has this contradiction. We will fix it of the v0.37.x version that we are working on. What the spec should state everywhere is that ADDED does not add the TX to the mempool.

Reply:
So after reading https://github.com/tendermint/tendermint/issues/9269, I realized the approach to setting actions is much simpler and we would avoid having to deal with any actions at all!

Reply:
I think we shouldn't assume that just yet. It seems like there is some consensus, but we should wait for additional stakeholders to weigh in @alexanderbez.

Reply:
Absolutely. We won't merge this PR/accept the ADR until that is resolved so we can make the necessary adjustments in the ADR 👍